File: frontend/types/game.ts
/**
 * Game-related TypeScript types
 */

export type RouteMode = 'shortest' | 'explorer';

export interface Country {
  id: string;
  code: string;
  name_ar: string;
  name_en: string;
  flag_emoji: string | null;
}

export interface CountryFull extends Country {
  continent: string | null;
  region: string | null;
  population: number | null;
  area_km2: number | null;
  capital_ar: string | null;
  capital_en: string | null;
}

export interface UserProgress {
  guesses: GuessEntry[];
  hints_used: number;
  completed: boolean;
  score: number | null;
}

export interface DailyChallenge {
  id: string;
  challenge_date: string;
  start_country: Country;
  end_country: Country;
  shortest_path: number;
  mode: RouteMode;
  path_country_codes: string[];
  user_progress: UserProgress | null;
}

export interface GuessEntry {
  country_id: string;
  country_code: string;
  name_ar: string;
  name_en?: string;
  flag_emoji: string | null;
  emoji: ScoreEmoji;
  order: number;
}

export type ScoreEmoji = '🟢' | '🟡' | '🟠' | '🔴' | '⚫';

export interface GuessRequest {
  challenge_id: string;
  country_id: string;
  mode: RouteMode;
}

export interface GuessResponse {
  country: Country;
  score_emoji: ScoreEmoji;
  score_description: string;
  is_on_shortest_path: boolean;
  is_destination: boolean;
  game_complete: boolean;
  total_guesses: number;
  score: number | null;
  route_mode: RouteMode;
  gap_from_optimal: number | null;
  quality_tier: 'perfect' | 'near_optimal' | 'good_discovery' | 'scenic' | null;
  quality_explanation_ar: string | null;
}

export interface HintRequest {
  challenge_id: string;
  mode: RouteMode;
}

export interface HintResponse {
  hint_type: 'progressive_1' | 'progressive_2' | 'progressive_3';
  hint_data: Record<string, unknown>;
  hints_remaining: number;
}

export interface GameStats {
  games_played: number;
  games_won: number;
  win_rate: number;
  current_streak: number;
  max_streak: number;
  average_guesses: number;
  hints_used_total: number;
  last_played: string | null;
}

export interface PracticeSessionRequest {
  start_country_id: string;
  end_country_id: string;
  mode: RouteMode;
}

export interface PracticeSession {
  session_id: string;
  mode: 'practice';
  route_mode: RouteMode;
  start_country: Country;
  end_country: Country;
  shortest_path: number;
  path_country_codes: string[];
  user_progress: UserProgress | null;
}

export interface PracticeGuessRequest {
  session_id: string;
  country_id: string;
  mode: RouteMode;
}

export interface PracticeHintRequest {
  session_id: string;
  mode: RouteMode;
}

export interface GameCompleteData {
  completed: boolean;
  score: number;
  total_guesses: number;
  hints_used: number;
  shortest_path: number;
  is_optimal: boolean;
}

// Score emoji descriptions in Arabic
export const EMOJI_DESCRIPTIONS: Record<ScoreEmoji, string> = {
  '🟢': 'ممتاز',
  '🟡': 'جيد',
  '🟠': 'مقبول',
  '🔴': 'بعيد',
  '⚫': 'قارة مختلفة',
};

File: backend/app/schemas/game.py
"""
Game-related Pydantic schemas for daily challenges.
"""

from datetime import date, datetime
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, Field

from app.schemas.country import CountryBrief

RouteMode = Literal["shortest", "explorer"]


class UserProgress(BaseModel):
    """User's progress on a challenge."""

    guesses: list[dict[str, Any]]
    hints_used: int
    completed: bool
    score: int | None = None


class DailyChallengeResponse(BaseModel):
    """Daily challenge response."""

    id: UUID
    challenge_date: date
    start_country: CountryBrief
    end_country: CountryBrief
    shortest_path: int
    mode: RouteMode = "shortest"
    path_country_codes: list[str] = Field(default_factory=list)
    user_progress: UserProgress | None = None

    class Config:
        from_attributes = True


class GuessRequest(BaseModel):
    """Request for submitting a guess."""

    challenge_id: UUID
    country_id: UUID
    mode: RouteMode = "shortest"


class GuessEntry(BaseModel):
    """Single guess entry in history."""

    country_id: UUID
    name_ar: str
    flag_emoji: str | None
    emoji: Literal["🟢", "🟡", "🟠", "🔴", "⚫"]
    order: int


class GuessResponse(BaseModel):
    """Response after submitting a guess."""

    country: CountryBrief
    score_emoji: Literal["🟢", "🟡", "🟠", "🔴", "⚫"]
    score_description: str
    is_on_shortest_path: bool
    is_destination: bool
    game_complete: bool
    total_guesses: int
    score: int | None = None
    route_mode: RouteMode = "shortest"
    gap_from_optimal: int | None = None
    quality_tier: Literal["perfect", "near_optimal", "good_discovery", "scenic"] | None = None
    quality_explanation_ar: str | None = None


class HintRequest(BaseModel):
    """Request for using a hint."""

    challenge_id: UUID
    mode: RouteMode = "shortest"


class HintResponse(BaseModel):
    """Response with hint data."""

    hint_type: Literal["progressive_1", "progressive_2", "progressive_3"]
    hint_data: dict[str, Any]
    hints_remaining: int


class GameCompleteResponse(BaseModel):
    """Response when game is completed."""

    completed: bool
    score: int
    total_guesses: int
    hints_used: int
    shortest_path: int
    is_optimal: bool  # True if user found shortest path


class GameStatsResponse(BaseModel):
    """User's game statistics."""

    games_played: int
    games_won: int
    win_rate: float
    current_streak: int
    max_streak: int
    average_guesses: float
    hints_used_total: int
    last_played: date | None


class PracticeSessionCreateRequest(BaseModel):
    """Request to start a practice session."""

    start_country_id: UUID
    end_country_id: UUID
    mode: RouteMode = "shortest"


class PracticeSessionResponse(BaseModel):
    """Practice session response."""

    session_id: UUID
    mode: Literal["practice"] = "practice"
    route_mode: RouteMode = "shortest"
    start_country: CountryBrief
    end_country: CountryBrief
    shortest_path: int
    path_country_codes: list[str] = Field(default_factory=list)
    user_progress: UserProgress | None = None


class PracticeGuessRequest(BaseModel):
    """Request for submitting a practice guess."""

    session_id: UUID
    country_id: UUID
    mode: RouteMode = "shortest"


class PracticeHintRequest(BaseModel):
    """Request for using a hint in practice mode."""

    session_id: UUID
    mode: RouteMode = "shortest"
